File: invoice-app-backend/src/main/java/com/invoiceapp/backend/invoice/domain/InvoiceStatus.java
package com.invoiceapp.backend.invoice.domain;

import java.util.Set;

public enum InvoiceStatus {
    DRAFT,
    SENT,
    PAID,
    OVERDUE,
    CANCELLED;

    public Set<InvoiceStatus> allowedTransitions() {
        return switch (this) {
            case DRAFT     -> Set.of(SENT, CANCELLED);
            case SENT      -> Set.of(PAID, OVERDUE, CANCELLED);
            case OVERDUE   -> Set.of(PAID, CANCELLED);
            case PAID, CANCELLED -> Set.of();
        };
    }

    public boolean canTransitionTo(InvoiceStatus target) {
        return allowedTransitions().contains(target);
    }
}

File: invoice-app-frontend/src/utils/invoiceStatus.ts
import type { InvoiceStatus } from "@/types";

export const statusVariant: Record<InvoiceStatus, 'default' | 'secondary' | 'destructive' | 'outline'> = {
    DRAFT: 'secondary',
    SENT: 'default',
    PAID: 'outline',
    OVERDUE: 'destructive',
    CANCELLED: 'secondary'
};

export const statusLabel: Record<InvoiceStatus, string> = {
    DRAFT: 'Draft',
    SENT: 'Sent',
    PAID: 'Paid',
    OVERDUE: 'Overdue',
    CANCELLED: 'Cancelled'
};